File: packages/prebid_mobile_sdk_gam/android/src/main/kotlin/com/prebid/prebid_mobile_sdk_gam/GamBannerAdViewFactory.kt
package com.prebid.prebid_mobile_sdk_gam

import android.content.Context
import android.view.View
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import org.prebid.mobile.AdSize
import org.prebid.mobile.api.exceptions.AdException
import org.prebid.mobile.api.rendering.BannerView
import org.prebid.mobile.api.rendering.listeners.BannerViewListener
import org.prebid.mobile.eventhandlers.GamBannerEventHandler

/// PlatformView factory for GAM-rendered banners. Mirrors the core
/// BannerAdViewFactory but builds the BannerView with a [GamBannerEventHandler]
/// so Google Ad Manager renders the ad.
class GamBannerAdViewFactory(
    private val messenger: BinaryMessenger,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val params = args as? Map<*, *> ?: emptyMap<String, Any>()
        return GamBannerPlatformView(context, viewId, messenger, params)
    }
}

class GamBannerPlatformView(
    context: Context,
    viewId: Int,
    messenger: BinaryMessenger,
    params: Map<*, *>,
) : PlatformView {

    private val bannerView: BannerView
    private val methodChannel: MethodChannel

    init {
        val configId = params["configId"] as? String ?: ""
        val gamAdUnitId = params["gamAdUnitId"] as? String ?: ""
        val width = params["width"] as? Int ?: 320
        val height = params["height"] as? Int ?: 50
        val isVideo = params["isVideo"] as? Boolean ?: false
        val autoLoad = params["autoLoad"] as? Boolean ?: true
        val refreshInterval = params["refreshIntervalSeconds"] as? Int

        methodChannel = MethodChannel(messenger, "prebid_mobile_sdk_gam/banner_$viewId")

        val eventHandler = GamBannerEventHandler(context, gamAdUnitId, AdSize(width, height))
        bannerView = BannerView(context, configId, eventHandler)

        if (isVideo) {
            bannerView.videoPlacementType = org.prebid.mobile.api.data.VideoPlacementType.IN_BANNER
        }

        if (refreshInterval != null && refreshInterval > 0) {
            bannerView.setAutoRefreshDelay(refreshInterval)
        }

        bannerView.setBannerListener(object : BannerViewListener {
            override fun onAdLoaded(view: BannerView) {
                view.bidResponse?.winningBid?.let { bid ->
                    methodChannel.invokeMethod(
                        "onAdSize",
                        mapOf(
                            "width" to bid.width.toDouble(),
                            "height" to bid.height.toDouble(),
                        ),
                    )
                }
                methodChannel.invokeMethod("onAdLoaded", null)
            }

            override fun onAdDisplayed(view: BannerView) {
                methodChannel.invokeMethod("onAdDisplayed", null)
            }

            override fun onAdFailed(view: BannerView, exception: AdException?) {
                methodChannel.invokeMethod("onAdFailed", exception?.message ?: "Unknown error")
            }

            override fun onAdClicked(view: BannerView) {
                methodChannel.invokeMethod("onAdClicked", null)
            }

            override fun onAdClosed(view: BannerView) {
                methodChannel.invokeMethod("onAdClosed", null)
            }
        })

        if (autoLoad) {
            bannerView.loadAd()
        }
    }

    override fun getView(): View = bannerView

    override fun dispose() {
        bannerView.destroy()
    }
}

File: packages/prebid_mobile_sdk_gam/ios/prebid_mobile_sdk_gam/Sources/prebid_mobile_sdk_gam/GamBannerAdViewFactory.swift
import Flutter
import UIKit
import GoogleMobileAds
import PrebidMobile
import PrebidMobileGAMEventHandlers

/// PlatformView factory for GAM-rendered banners. Mirrors the core
/// BannerAdViewFactory but builds the BannerView with a `GAMBannerEventHandler`
/// so Google Ad Manager renders the ad.
class GamBannerAdViewFactory: NSObject, FlutterPlatformViewFactory {

    private let messenger: FlutterBinaryMessenger

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        return GamBannerPlatformView(
            frame: frame,
            viewId: viewId,
            messenger: messenger,
            args: args as? [String: Any] ?? [:]
        )
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

class GamBannerPlatformView: NSObject, FlutterPlatformView, PrebidMobile.BannerViewDelegate {

    private let bannerView: PrebidMobile.BannerView
    private let methodChannel: FlutterMethodChannel

    init(
        frame: CGRect,
        viewId: Int64,
        messenger: FlutterBinaryMessenger,
        args: [String: Any]
    ) {
        let configId = args["configId"] as? String ?? ""
        let gamAdUnitId = args["gamAdUnitId"] as? String ?? ""
        let width = args["width"] as? Int ?? 320
        let height = args["height"] as? Int ?? 50
        let isVideo = args["isVideo"] as? Bool ?? false
        let autoLoad = args["autoLoad"] as? Bool ?? true
        let refreshInterval = args["refreshIntervalSeconds"] as? Int

        let adSize = CGSize(width: width, height: height)

        methodChannel = FlutterMethodChannel(
            name: "prebid_mobile_sdk_gam/banner_\(viewId)",
            binaryMessenger: messenger
        )

        let eventHandler = GAMBannerEventHandler(
            adUnitID: gamAdUnitId,
            validGADAdSizes: [nsValue(for: adSizeFor(cgSize: adSize))]
        )
        bannerView = PrebidMobile.BannerView(
            frame: CGRect(origin: .zero, size: adSize),
            configID: configId,
            adSize: adSize,
            eventHandler: eventHandler
        )

        super.init()

        if isVideo {
            bannerView.adFormat = .video
        }

        if let interval = refreshInterval, interval > 0 {
            bannerView.refreshInterval = TimeInterval(interval)
        }

        bannerView.delegate = self

        if autoLoad {
            bannerView.loadAd()
        }
    }

    func view() -> UIView {
        return bannerView
    }

    // MARK: - BannerViewDelegate

    func bannerViewPresentationController() -> UIViewController? {
        return UIApplication.shared.keyWindow?.rootViewController
    }

    func bannerView(_ bannerView: PrebidMobile.BannerView, didReceiveAdWithAdSize adSize: CGSize) {
        methodChannel.invokeMethod("onAdSize", arguments: [
            "width": Double(adSize.width),
            "height": Double(adSize.height),
        ])
        // iOS reports load and render as one event; Android splits them into
        // onAdLoaded + onAdDisplayed, so emit both here for cross-platform parity.
        methodChannel.invokeMethod("onAdLoaded", arguments: nil)
        methodChannel.invokeMethod("onAdDisplayed", arguments: nil)
    }

    func bannerView(_ bannerView: PrebidMobile.BannerView, didFailToReceiveAdWith error: Error) {
        methodChannel.invokeMethod("onAdFailed", arguments: error.localizedDescription)
    }

    func bannerViewWillPresentModal(_ bannerView: PrebidMobile.BannerView) {
        methodChannel.invokeMethod("onAdClicked", arguments: nil)
    }

    func bannerViewDidDismissModal(_ bannerView: PrebidMobile.BannerView) {
        methodChannel.invokeMethod("onAdClosed", arguments: nil)
    }
}
